sql: support valargs for bin ops

madelynnblue opened the pull request:
This adds support for any binary operation where both sides can be the
same type (like numbers). It doesn't work for dates, since date addition
doesn't make sense.

Fixes #3628

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3724)

<!-- Reviewable:end -->

tamird commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/pgwire_test.go, line 216 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-K7fOy3ViTAwdfrEb9MD:-K7fOy3ViTAwdfrEb9ME:805635601) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26b9ff181d3192cfc079745ce5ad5b5d431ae44/sql/pgwire_test.go#L216)):*
how about something like `SELECT 1 + $1, $1 + $2`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/pgwire_test.go, line 286 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-K7l7b3ANS9vEv4o_qhK:-K7l7b3ANS9vEv4o_qhL:2007823731) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26b9ff181d3192cfc079745ce5ad5b5d431ae44/sql/pgwire_test.go#L286)):*
`valarg` isn't a very descriptive term to see in an error message. Perhaps `s/valarg/param/g` or `s/valarg/placeholder/g`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/pgwire_test.go, line 216 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-K7fOy3ViTAwdfrEb9MD:-K7lwFcztqXUyKgW4gN7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26b9ff181d3192cfc079745ce5ad5b5d431ae44/sql/pgwire_test.go#L216)):*
Done.

---

*[sql/pgwire_test.go, line 286 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-K7l7b3ANS9vEv4o_qhK:-K7lw_UQi-p5Vdj0xwEz:460066025) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26b9ff181d3192cfc079745ce5ad5b5d431ae44/sql/pgwire_test.go#L286)):*
That will get done with https://github.com/cockroachdb/cockroach/pull/3545. I think it can stay out of this PR because it would require a larger change that has nothing to do with binary ops.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/type_check.go, line 67 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-K7m2LDDf2UZOoap5a7v:-K7m2LDE5VyQgm05DxSh:531283558) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a1684b3ca7d59c32ed5154e9b0a3d71631ecd0a/sql/parser/type_check.go#L67)):*
i expected (when i posted my test case suggestion) that this would have to change to two separate `if` clauses, rather than the current `else if`. Can you figure out why this works? o.O

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: support valargs for bin ops

This adds support for any binary operation where both sides can be the
same type (like numbers). It doesn't work for dates, since date addition
doesn't make sense.

Fixes #3628

mjibson commented:
Fixed so tests should pass now.

---

Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/parser/type_check.go, line 67 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-K7m2LDDf2UZOoap5a7v:-K7muOfX1wdre9SFEL8P:985717685) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a1684b3ca7d59c32ed5154e9b0a3d71631ecd0a/sql/parser/type_check.go#L67)):*
As discussed, but for the archives: this works because `$1` in the addition has a `TypeCheck()` result as a DInt, not a DValArg, because we already know what its type will be.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-:-K7mubLUotP4z-jelGWt:-1834564724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/pgwire_test.go, line 286 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724#-K7l7b3ANS9vEv4o_qhK:-K80kERIagrQdaamHWS5:764053012) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26b9ff181d3192cfc079745ce5ad5b5d431ae44/sql/pgwire_test.go#L286)):*
Resurrecting the dead here, but #3545 is closing. What's the plan there?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3724)*
<!-- Sent from Reviewable.io -->
